File: cs50x/pset1/greedy.c
/****************************************************************************
 * greedy.c
 *
 * The objective of this application is  to use the greedy algorithm to display
 * the least amount of coins that can be used in USD for change owed.
 * 
 * Usage        : ./greedy
 * Description  : http://docs.cs50.net/problems/greedy/greedy.html
 * Testing      : check50 2016.greedy greedy.c
 ***************************************************************************/

#include <stdio.h>
#include <cs50.h>

int main( void ) {

    float num; // Input Amount of money
    int left; // How much money left?

    do {
        printf( "O hai! How much change is owed?\n" );
        num = GetDouble();
    } while ( num < 0.0 );

    // Well, its a way to convert float to to int.
    left = (int) ( num * 1000 ) / 10;

    int coins[4] = { 25, 10, 5, 1 };
    int result   = 0; // Storing coins number here.

    // Going in for loop
    for ( int i = 0; i < sizeof( coins ) / sizeof ( int ) ; i++ ) {
        result += left / coins[i]; // so how many coins do we need for this?
        left = left % coins[i]; // adn assigning to left what is left after division
    }

    printf( "%i\n", result ); // output result.
}

File: HarvardX_CS50x/pset6/greedy.py
'''
	greedy.py

	The objective of this application is  to use the greedy algorithm to display
	the least amount of coins that can be used in USD for change owed.

	Usage        : python3 greedy.py
	Description  : http://docs.cs50.net/problems/mario/more/mario.html
'''

from re import search
import sys

# Actually clone of get_float by cs50 team.
def get_float( StringOfFloat ) :
	if StringOfFloat is None or len(StringOfFloat) is 0 :
		return False
	elif search( r"^[+-]?\d*(?:\.\d*)?$", StringOfFloat ) :
		try:
			return float( StringOfFloat )
		except ValueError:
			pass
	return False

def main():
	# Initial value for coins storage. In False.
	Coins = False;

	# Getting Value from User.
	while  Coins is False or Coins <= 0.0 :
		CoinsStr = input( "O hai! How much change is owed?\n" )
		Coins = get_float( CoinsStr )

	# Bring value to Int
	Coins = int( Coins * 100 );

	# List of coins amounts.
	CoinsAmount = [ 25, 10, 5, 1 ];
	# TOtal value init value.
	TotalCoins = 0;

	# Do math.
	for coinAmount in CoinsAmount :
		TotalCoins +=  Coins // coinAmount;
		Coins %= coinAmount;
		#print(Coins, TotalCoins);

	# Print value.
	print( int( TotalCoins ) );


if __name__ == '__main__':
	main();
